### ISSUE TITLE
Gap junction implementation not strictly C++-standard compliant

### ISSUE THREAD

**Issue by heplesser** [nest/nest-simulator#181]
Compiling NEST with Clang reveals a conflict with C++ standard compliance in the implementation of gap junctions:
```
In file included from ../../src/nestkernel/sibling_container.cpp:23:
../../src/nestkernel/event.h:770:10: error: calling a private constructor of class 'std::__1::__wrap_iter<unsigned int *>'
   fwit iit( reinterpret_cast< uint_t* >( &( *rhs ) ) );
```
The problem here is that we attempt to initialize an iterator with a pointer. This is not standard-compliant, since STL implementations may choose to implement vector iterators with data types other than pointers.

The reason for the construction above is that we need to exchange numerical information (doubles) as unsigned long, and want to do so without copying data (for efficiency). Currently, we do this by having a vector of doubles and coercing vector<ulong> iterators onto that vector.

I have [written a little example program](https://gist.github.com/heplesser/d5a1b2e0067a817168a6#file-union_vector-cpp) that uses a vector of a double-ulong union to achieve the same effect. Maybe this can be used to implement the double/ulong duality in a standard-compliant way.

@janhahne @apeyser @tammoippen @jougs I hope you also consider this a neat little technical problem in C++.

**Comment by janhahne** [comment]
@heplesser When @mhelias and I build the `CoeffArrayIterator` class we were aware, that the `std::vector<double>::iterator` <-> `std::vector<uint>::iterator` transformation using pointers and `reinterpret_cast` might be critical, but then it worked just fine with gcc and also on JuQueen and the K computer.

I agree that we should look for a more standard-compliant way to do this, but we should also take care that we do not lose too much performance on this. Otherwise it would be debatable if it is benifical to change this.

If I understand your little example program correctly this would mean:
- the data which needs to be send in `hh_psc_alpha_gap` is written into a vector of type DUVector
- the `GapJEvent` saves an `DUVector::ulong_iterator` for begin and end
- the next step would be the `std::copy` into the `secondary_event_buffer` which is of type `std::vector<uint>` 
```cpp
inline fwit& GapJEvent::operator>>( fwit& pos )
{
  input_stream( synid_, pos );
  input_stream( sender_gid_, pos );
  std::copy( begin().pos_, end().pos_, pos );

  return pos;
}
```
but at this point aren't we still having the same problem as before, as we need to convert an `std::vector<Elem>::iterator` to an `std::vector<uint>::iterator`? Same for the transformation later on the other side.

**Comment by heplesser** [comment]
@janhahne `std::copy()` allows different iterator types for source and destination. I won't have time to test this before noon, unfortunately.

**Comment by apeyser** [comment]
I'm not sure I understand what's going on here. Is an array of doubles being interpreted as uint by bit-casting? Why? And why is this performance critical? Are doubles & ints being stuffed into the same array, and are they being distinguished some how? Or is this an attempt to override a double array with integer? And if so, why is a possibly varying bit length type (unsigned int) being traded with a fixed length type (ieee floats)?

**Comment by janhahne** [comment]
@apeyser Here is a more detailed description of what is going on there

In `hh_psc_alpha_gap` a `GapJEvent` is filled with data:
```cpp
std::vector< double_t > new_coefficients( quantity, 0.0 );

[...]

// Send gap-event
GapJEvent ge;
ge.set_coeffarray( new_coefficients );
network()->send_secondary( *this, ge );
```

Later this is written into the `std::vector< std::vector< uint_t > > secondary_events_buffer_;`: via `inline fwit& GapJEvent::operator>>( fwit& pos )`:
```cpp
inline void
Scheduler::send_remote( thread t, SecondaryEvent& e )
{
  // put the secondary events in a buffer for the remote machines
  size_t old_size = secondary_events_buffer_[ t ].size();

  secondary_events_buffer_[ t ].resize( old_size + e.size() );
  fwit it = secondary_events_buffer_[ t ].begin() + old_size;
  e >> it;
}
```

which is attached to `std::vector< uint_t > local_grid_spikes_;` in `nest::Scheduler::collocate_buffers_` and used for the MPI-communication via `Communicator::communicate( local_grid_spikes_, global_grid_spikes_, displacements_ );`

On the receiving side in `nest::Scheduler::deliver_events_` the `GapJEvent` is read from `std::vector< uint_t > global_grid_spikes_;` via `inline fwit& GapJEvent::operator<<( fwit& pos )`:
```cpp
fwit readpos = global_grid_spikes_.begin() + pos[ pid ];

      while ( true )
      {
        // we must not use uint_t for the type, otherwise
        // the encoding will be different on JUQUEEN for the
        // index written into the buffer and read out of it
        synindex synid;
        output_stream( synid, readpos );

        if ( synid == invalid_synindex )
          break;
        --readpos;

        net_->connection_manager_.assert_valid_syn_id( synid );

        *( secondary_events_prototypes_[ t ][ synid ] ) << readpos;

        net_->connection_manager_.send_secondary(
          t, *( secondary_events_prototypes_[ t ][ synid ] ) );
      } // of while (true)
```

and send to `hh_psc_alpha_gap` via `nest::hh_psc_alpha_gap::handle( GapJEvent& e )`

_______________________

In our original implementation `GapJEvent` had a member `std::vector< double_t > coeff_array` and the workflow was as follows:
- the double vector from `hh_psc_alpha_gap` was copied to the double vector in `GapJEvent`
- the double vector from the `GapJEvent` was copied to the uint vector  `secondary_events_buffer_`
and on the other side:
- the uint vector `global_grid_spikes_` was copied to the double vector in `GapJEvent`

The many copies were however obviously bad for memory usage and time performance. So we introduced the `class CoeffArrayIterator`. Now `GapJEvent` has no double array, but iterators pointing to the original memory location of the array (meaning the vector in `hh_psc_alpha_gap` on the sending side and the `global_grid_spikes` vector on the receiving side).

The thing here is that we needed to convert `std::vector<double>::iterator` into `std::vector<uint>::iterator` and vice versa. That is done using pointers and `reinterpret_cast`, which is causing the problem here.

**Comment by apeyser** [comment]
So, if I understand correctly, you are trying to override a vector of uint's with a vector of doubles, hoping that all the vector of uint machinery will work properly in between including remote passing, and then cast the whole thing back to the correct type at the end?

This won't work in 32 bit environments where sizeof(uint) != sizeof(double), for starters. Second, even if we assume a 64 bit environment or a 32 bit environment where integers == longs, you want to reinterpret the underlying ints <-> double, and not the enclosing vector where anything can happen. So instead of vector<double>, you should use an adaptor iterator class that does your casts so that you know you're not doing anything illegal under the hood, just risking an undefined behavior when you're in the wrong environment by chopping off the ends of your doubles or filling them with random bits.

A better solution is to make the remote send and receive opaque types of 64 bits, and force the sender and receiver to fill them properly using a simpler variation on the underlying union type mentioned by @heplesser . You could make a class that is a vector<union> + a type tag that is checked in asserts. You wouldn't even need the full reimplements of the vector interface like hep, did, just make it a public base class. If you just do it->double_val or it->int_val, you'd have no cost. Of course, you could make an inline cast, or the inline iterator adaptors, but I don't see that it gives any added type safety once you've already started using a dynamic union while simultaneously making the code opaque.

So, to summarize: what you want is to cast between an array of doubles and unsigned integers. Better choices are: reinterpret cast between double and uint, and then move the uint in and out of the array (still pretty bad); or have an array of union types, and pull out the right union member.

**Comment by apeyser** [comment]
Do we have some measures or theoretical model for the performance loss for copying into a vector<uint> of fixed size? And that still would leave the issue that bitcasting between double and int isn't 32 bit safe --- there would need to be a configuration flag to turn off gap junction for 32 bit machines and a compile time check for the sizeof(unsigned int) >= sizeof(double) if gap junction code is included.

**Comment by janhahne** [comment]
@apeyser We do not at all rely on sizeof(uint) == sizeof(double). I just tried to describe the general problem why we used the `CoeffArrayIterator` which causes the compiler error with clang and the process of the data. For the details I suggest you have a look at the actual code in `event.h` (lines 913-975), which is pretty much self explaining.

I do not doubt that there are smarter or cleaner ways to do this and I am happy for any idea on that, but the current version works fine regardless of the size of `unsigned int`  and `double` on the used machine.

**Comment by apeyser** [comment]
Okay, so then I wasn't understanding correctly. So is the problem is that you don't want to copy a uint -> a uint before using operator>> which does copy a uint -> a uint? And that's needed to stuff the values through send remote using a virtual >> operator?

So, I see know that you've defined a post-incrementer that then scales the increment in case sizeof(double) is a multiple of sizeof(uint), assuming that the vector internals are packed so that one double can overlap two uints internally. I had missed that -- I think I see now what you were doing.

Even without problems from non-flat array implementations for vectors, the compiler is free to do whatever it wants when you've done this sort of aliasing. Even if you did in straight C with arrays so that you can guarantee that the bit layout that you assume is correct, an aggressive optimization could do almost anything.

Why not remove most of the code, and instead of trying to coerce the dfwit into an fwit, put a loop inside your GapJEvent::operator>> copying the bits from the double iterator into the output uint iterator? WIth a compile time check, you could then split doubles into high & low bits, or just directly ship the bits. If I'm understanding correctly (and given that I haven't understood twice what the goal of the operation is, I'm not sure that I do), this is the one and only use of the CoeffArrayIterator machinery -- to be used inside of operator>> and operator<< to copy, which are virtual methods called insided of send_remote & deliver_events.

Also, the inlining seems to conflict with the virtual definition. Where I see that those two operators get called (I could be wrong), is on references and pointers to SecondaryEvents, in which the operators <<|>> are defined as abstract virtual -- so they can't be inlined.

So what I'm suggesting is along the lines of first defining:
```c++
vector<double>::iterator GapJEvent::diit_begin, diit_end;
```
and then inside a compile time check, either:
```c++
    virtual fwit& GapJEvent::operator>>(fwit& pos) {
      for (i = diit_begin(); i != diit_end(); i++) {
        double d = *i;
        *pos++ = *reinterpret_cast<uint_t*>(&i);
      }
      return pos;
    }
```
for the 64 bit case, or
```c++
    virtual fwit& GapJEvent::operator>>(fwit& pos) {
      for (i = diit_begin(); i != diit_end(); i++) {
        double d = *i;
        char* c = reinterpret_cast<char*>(&d);
        *pos++ = reinterpret_cast<uint_t*>(c);
        *pos++ = reinterpret_cast<uint_t*>(c+sizeof(uint_t));
      }
      return pos;
    }
```
for the 32 bit case. 

For the reverse case, store a vector<uint_t>::iterator, and then put a little code to do the reverse operation when handle tries to walk over the vector.

Then most of the rest can go away... everything is perfectly legal and much shorter.

**Comment by janhahne** [comment]
@apeyser I think now we are talking about the same thing! Thank you for the valuable comments. I created a branch ( https://github.com/janhahne/nest-simulator/tree/gap_fix181 ) which removes the `CoeffArrayIterator` and instead uses `std::vector` iterators in `GapJEvent` as you suggested. For starters this should fix the compiler error mentioned in this issue. You are right that the operator-functions and the `GapJEvent::size()`-function are always called from `*SecondaryEvent`. I hence removed the `inline` in the fix. Would you have a look at this?

Furthermore if I understand you correctly there is also another problem with our implementation, which is that `&(*pos)` and `&(*(pos+1))` are not necessarily located next to each other in the memory which means, that when we do an operation like
```cpp
template < typename T >
void
input_stream( T d, fwit& pos )
{
  memcpy( &( *pos ), &d, sizeof( d ) );
  pos += size_uint_t( d );
}
```

and `size_uint_t( d )` is >= 2 we could possibly overwrite the wrong part of memory? On my local machine, JuQueen and K this has not been a problem so far, but I see why this might be critical.

What do you think about a fix like
```cpp
template < typename T >
void
input_stream( T d, fwit& pos )
{
  size_t size_uint = size_uint_t( d );
  for( int i=0; i < size_uint; i++ )
     memcpy( &( *( pos+i ) ), &d + i*sizeof( uint_t ), sizeof( uint_t ) );
  pos += size_uint;
}
```
This could if I am not mistaking prevent that one needs different versions for 32 and 64 bit? On the other hand of course one is doing a lot of function calls to `size_uint_t`. Maybe this could be done at compile time...

**Comment by heplesser** [comment]
@janhahne @apeyser I would suggest that we do not attempt to solve problems before they actually occur, especially if they seem quite unlikely, such as non-contiguous storage of vector elements. The important thing is to have solid tests in place that would flag any trouble in case storage is not contiguous. `test_gap_junction.sli` at present only checks that gap junctions can be created, while `test_gap_junction_mpi.sli` tests the consistency when four VPs are distributed across 1, 2, and 4 MPI processes, but no test checks against a "ground truth", i.e., a result we know to be correct. We need to add such a test. The test could build on the MPI-test, but
- should also be included in the non-mpi test, so we get the check when building without MPI
- needs to be included in the MPI test so we detect trouble introduced by MPI communication
- should look at membrane potential in addition to spike times for greater sensitivity (we have a `ToUnitTestPrecision` for comparison of doubles with a given number of decimals)

**Comment by heplesser** [comment]
@janhahne While reviewing the gap junction code, I found it difficult to keep a good overview over the different data types and data members due to some rather cryptic names. Longer names generally make for better readability in my experience. Would you consider the following:
- `GapJEvent`-> `GapJunctionEvent`
- `fwit` -> `std::vector< uint_t >::iterator` 
- `dfwit` -> `std::vector< double_t >::iterator`
- `dit_begin_` -> `coeffarray_as_doubles_begin_`
- `uiit_begin_` -> `coeffarray_as_uints_begin_`
and corresponding for `*end_`?
We will read this code only very rarely, so when we read it again a year from now or so, hardly anyone will remember the meaning of `fwit` or `dit_begin_`, requiring a lot of look-up to be able to read the code. More verbose names will make reading much more straightforward.

**Comment by janhahne** [comment]
@heplesser The suggested name changes sound reasonable. I will include them in PR #181.

There actually is a third test in the testsuite doing what you ask for. It is hiding under the name of the neuron model `test_hh_psc_alpha_gap`. The only thing missing is that it is not included in the MPI-test.

**Comment by apeyser** [comment]
@janhahne: Looks good, except the &d needs a char* cast, but the concept is sound, and to help the compiler with some const's for the bounds. I made a suggestion on the PR on how to turn the size_uint call into something compile-time collapsible to a const.

I'd like to think about later (after merge), moving some of this code into a separate file where we can make this reusable, since the problem is general and I know at least one other project that would like to use this.

@heplesser: undefined behavior is very tricky to debug -- it's not just the non-contiguous vector, but compiler optimizations that could drag us into long debugs. 

A summary of this whole thing as a comment would very helpful, maybe as pseudo-code, which would also help with the name issue.

**Comment by heplesser** [comment]
@janhahne I think we have enough tests then.
